src/main/java/xyz/nucleoid/plasmid/game/player/LocalJoinOffer.java

```diff
@@ -0,0 +1,83 @@
+package xyz.nucleoid.plasmid.game.player;
+
+import com.mojang.authlib.GameProfile;
+import net.minecraft.entity.player.PlayerEntity;
+import net.minecraft.server.network.ServerPlayerEntity;
+import net.minecraft.server.world.ServerWorld;
+import net.minecraft.text.Text;
+import net.minecraft.util.math.Vec3d;
+import net.minecraft.world.GameMode;
+
+import java.util.*;
+import java.util.function.Consumer;
+
+public record LocalJoinOffer(Collection<ServerPlayerEntity> serverPlayers, JoinIntent intent) implements JoinOffer {
+    @Override
+    public Collection<GameProfile> players() {
+        return this.serverPlayers
+                .stream()
+                .map(PlayerEntity::getGameProfile)
+                .toList();
+    }
+
+    @Override
+    public Collection<UUID> playerIds() {
+        return this.serverPlayers
+                .stream()
+                .map(player -> player.getGameProfile().getId())
+                .toList();
+    }
+
+    @Override
+    public Collection<String> playerNames() {
+        return this.serverPlayers
+                .stream()
+                .map(player -> player.getGameProfile().getName())
+                .toList();
+    }
+
+    @Override
+    public JoinOfferResult.Accept accept(ServerWorld world, Vec3d position, float yaw, float pitch) {
+        return new Accept(world, position, yaw, pitch);
+    }
+
+    @Override
+    public JoinOfferResult.Reject reject(Text reason) {
+        return () -> reason;
+    }
+
+    public static class Accept implements JoinOfferResult.Accept {
+        private final ServerWorld world;
+        private final Vec3d position;
+        private final float yaw;
+        private final float pitch;
+
+        private final List<Consumer<PlayerSet>> thenRun = new ArrayList<>();
+
+        Accept(ServerWorld world, Vec3d position, float yaw, float pitch) {
+            this.world = world;
+            this.position = position;
+            this.yaw = yaw;
+            this.pitch = pitch;
+        }
+
+        @Override
+        public Accept thenRun(Consumer<PlayerSet> consumer) {
+            this.thenRun.add(consumer);
+            return this;
+        }
+
+        public void joinAll(PlayerSet players) {
+            for (var consumer : this.thenRun) {
+                consumer.accept(players);
+            }
+        }
+
+        public ServerWorld applyJoin(ServerPlayerEntity player) {
```

Comment:
I think there is some value still to running per-player logic in here, as this applies before the client is fully made aware of the teleport - so doing setup in here ensures that the state is finalised before the player sees the world again. Maybe the 'global' thenRun can still run after all players, but the per-player logic runs in here?

Reply:
In a different conversation we have reached a conclusion that the player is not available at this stage.